File: web/src/lib/agent-smith/parse.js
// =====================================================================
//  Agent Smith — parser.
//  Parses an Agent Smith script (the `agent {...}` / `society {...}`
//  grammar of the paper) into a specification AST. Pure, dependency-free.
//
//  Grammar (concrete surface):
//    agent <id> {
//      purpose (minimise <phi> | reach <outcome>)
//      scenes { scene <id> serves <purpose> with <hook> ... }
//      self { parts { a, b, ... } separations { (a, b: cost), ... } }
//      budget <num>
//      floor  <num>
//      coherence keeps { a, b, ... }
//    }
//    society <id> { <agent-or-ref> ... tie (a,b: cost) ... couple <K> }
//
//  The parser is deliberately forgiving about whitespace and comments
//  (// to end of line) and strict about structure, so that typing
//  (typecheck.js) — not parsing — is where an agent is rejected.
// =====================================================================

/** A parse error carries a message and a 1-based line number. */
export class ParseError extends Error {
  constructor(message, line) {
    super(line ? `line ${line}: ${message}` : message);
    this.line = line || null;
    this.name = "ParseError";
  }
}

// ---- tokenizer -------------------------------------------------------

const PUNCT = new Set(["{", "}", "(", ")", ",", ":", "="]);

function tokenize(src) {
  const toks = [];
  let line = 1;
  let i = 0;
  const n = src.length;
  while (i < n) {
    const c = src[i];
    if (c === "\n") {
      line++;
      i++;
      continue;
    }
    if (c === " " || c === "\t" || c === "\r") {
      i++;
      continue;
    }
    // line comment
    if (c === "/" && src[i + 1] === "/") {
      while (i < n && src[i] !== "\n") i++;
      continue;
    }
    if (PUNCT.has(c)) {
      toks.push({ t: c, line });
      i++;
      continue;
    }
    // string literal
    if (c === '"') {
      let j = i + 1;
      let s = "";
      while (j < n && src[j] !== '"') {
        s += src[j];
        j++;
      }
      if (j >= n) throw new ParseError("unterminated string", line);
      toks.push({ t: "str", v: s, line });
      i = j + 1;
      continue;
    }
    // number
    if (/[0-9]/.test(c) || (c === "-" && /[0-9]/.test(src[i + 1] || ""))) {
      let j = i + 1;
      while (j < n && /[0-9.eE+\-]/.test(src[j])) j++;
      const raw = src.slice(i, j);
      const v = Number(raw);
      if (Number.isNaN(v)) throw new ParseError(`bad number "${raw}"`, line);
      toks.push({ t: "num", v, line });
      i = j;
      continue;
    }
    // identifier / keyword
    if (/[A-Za-z_]/.test(c)) {
      let j = i + 1;
      while (j < n && /[A-Za-z0-9_.]/.test(src[j])) j++;
      toks.push({ t: "id", v: src.slice(i, j), line });
      i = j;
      continue;
    }
    throw new ParseError(`unexpected character "${c}"`, line);
  }
  toks.push({ t: "eof", line });
  return toks;
}

// ---- recursive-descent parser ---------------------------------------

class Parser {
  constructor(toks) {
    this.toks = toks;
    this.p = 0;
  }
  peek() {
    return this.toks[this.p];
  }
  next() {
    return this.toks[this.p++];
  }
  at(t) {
    return this.peek().t === t;
  }
  atKw(kw) {
    const tk = this.peek();
    return tk.t === "id" && tk.v === kw;
  }
  eat(t) {
    const tk = this.peek();
    if (tk.t !== t) throw new ParseError(`expected "${t}", got "${tk.v ?? tk.t}"`, tk.line);
    return this.next();
  }
  eatKw(kw) {
    const tk = this.peek();
    if (tk.t !== "id" || tk.v !== kw)
      throw new ParseError(`expected "${kw}", got "${tk.v ?? tk.t}"`, tk.line);
    return this.next();
  }
  id() {
    const tk = this.peek();
    if (tk.t !== "id") throw new ParseError(`expected identifier, got "${tk.v ?? tk.t}"`, tk.line);
    return this.next().v;
  }
  num() {
    const tk = this.peek();
    if (tk.t !== "num") throw new ParseError(`expected number, got "${tk.v ?? tk.t}"`, tk.line);
    return this.next().v;
  }

  // ---- top level ----
  parseProgram() {
    const tk = this.peek();
    if (this.atKw("agent")) return this.parseAgent();
    if (this.atKw("society")) return this.parseSociety();
    throw new ParseError(`expected "agent" or "society", got "${tk.v ?? tk.t}"`, tk.line);
  }

  // ---- agent ----
  parseAgent() {
    const line = this.peek().line;
    this.eatKw("agent");
    const name = this.id();
    this.eat("{");
    const spec = {
      kind: "agent",
      name,
      line,
      purpose: null,
      scenes: [],
      self: null,
      budget: null,
      floor: null,
      coherence: null,
    };
    while (!this.at("}")) {
      if (this.at("eof")) throw new ParseError('unclosed "agent" block', line);
      if (this.atKw("purpose")) spec.purpose = this.parsePurpose();
      else if (this.atKw("scenes")) spec.scenes = this.parseScenes();
      else if (this.atKw("self")) spec.self = this.parseSelf();
      else if (this.atKw("budget")) {
        this.eatKw("budget");
        spec.budget = this.num();
      } else if (this.atKw("floor")) {
        this.eatKw("floor");
        spec.floor = this.num();
      } else if (this.atKw("coherence")) spec.coherence = this.parseCoherence();
      else {
        const bad = this.peek();
        throw new ParseError(`unexpected "${bad.v ?? bad.t}" in agent body`, bad.line);
      }
    }
    this.eat("}");
    return spec;
  }

  parsePurpose() {
    this.eatKw("purpose");
    const tk = this.peek();
    if (this.atKw("minimise") || this.atKw("minimize")) {
      this.next();
      const potential = this.id();
      return { mode: "minimise", potential, line: tk.line };
    }
    if (this.atKw("reach")) {
      this.next();
      const outcome = this.id();
      return { mode: "reach", outcome, line: tk.line };
    }
    throw new ParseError(`purpose must be "minimise <phi>" or "reach <outcome>"`, tk.line);
  }

  parseScenes() {
    this.eatKw("scenes");
    this.eat("{");
    const scenes = [];
    while (!this.at("}")) {
      if (this.at("eof")) throw new ParseError('unclosed "scenes" block', this.peek().line);
      const line = this.peek().line;
      this.eatKw("scene");
      const name = this.id();
      this.eatKw("serves");
      const serves = this.id();
      this.eatKw("with");
      const hook = this.id();
      scenes.push({ name, serves, hook, line });
    }
    this.eat("}");
    return scenes;
  }

  parseSelf() {
    const line = this.peek().line;
    this.eatKw("self");
    this.eat("{");
    let parts = [];
    let separations = [];
    while (!this.at("}")) {
      if (this.at("eof")) throw new ParseError('unclosed "self" block', line);
      if (this.atKw("parts")) {
        this.eatKw("parts");
        parts = this.parseIdList();
      } else if (this.atKw("separations")) {
        this.eatKw("separations");
        separations = this.parseSeparations();
      } else {
        const bad = this.peek();
        throw new ParseError(`unexpected "${bad.v ?? bad.t}" in self body`, bad.line);
      }
    }
    this.eat("}");
    return { parts, separations, line };
  }

  // { a, b, c }
  parseIdList() {
    this.eat("{");
    const ids = [];
    while (!this.at("}")) {
      ids.push(this.id());
      if (this.at(",")) this.next();
      else if (!this.at("}")) throw new ParseError('expected "," or "}"', this.peek().line);
    }
    this.eat("}");
    return ids;
  }

  // { (a, b: 3), (b, c: 2), ... }
  parseSeparations() {
    this.eat("{");
    const seps = [];
    while (!this.at("}")) {
      const line = this.peek().line;
      this.eat("(");
      const a = this.id();
      this.eat(",");
      const b = this.id();
      this.eat(":");
      const cost = this.num();
      this.eat(")");
      seps.push({ a, b, cost, line });
      if (this.at(",")) this.next();
      else if (!this.at("}")) throw new ParseError('expected "," or "}"', this.peek().line);
    }
    this.eat("}");
    return seps;
  }

  parseCoherence() {
    const line = this.peek().line;
    this.eatKw("coherence");
    // optional "keeps"
    if (this.atKw("keeps")) this.next();
    const keeps = this.parseIdList();
    return { keeps, line };
  }

  // ---- society ----
  parseSociety() {
    const line = this.peek().line;
    this.eatKw("society");
    const name = this.id();
    this.eat("{");
    const spec = { kind: "society", name, line, members: [], ties: [], couple: null };
    while (!this.at("}")) {
      if (this.at("eof")) throw new ParseError('unclosed "society" block', line);
      if (this.atKw("agent")) spec.members.push(this.parseAgent());
      else if (this.atKw("tie")) {
        this.eatKw("tie");
        this.eat("(");
        const a = this.id();
        this.eat(",");
        const b = this.id();
        this.eat(":");
        const cost = this.num();
        this.eat(")");
        spec.ties.push({ a, b, cost });
      } else if (this.atKw("couple")) {
        this.eatKw("couple");
        spec.couple = this.num();
      } else {
        // a bare member reference by name
        const bad = this.peek();
        if (bad.t === "id") {
          spec.members.push({ kind: "ref", name: this.id(), line: bad.line });
        } else {
          throw new ParseError(`unexpected "${bad.v ?? bad.t}" in society body`, bad.line);
        }
      }
    }
    this.eat("}");
    return spec;
  }
}

/**
 * Parse an Agent Smith script into a spec AST.
 * @param {string} src
 * @returns {object} the spec (kind: "agent" | "society")
 * @throws {ParseError}
 */
export function parse(src) {
  const toks = tokenize(src);
  const parser = new Parser(toks);
  const spec = parser.parseProgram();
  if (!parser.at("eof")) {
    const tk = parser.peek();
    throw new ParseError(`unexpected trailing "${tk.v ?? tk.t}"`, tk.line);
  }
  return spec;
}

File: crates/agent-smith/src/parse.rs
// =====================================================================
//  Agent Smith — parser.
//  Parses an Agent Smith script into a specification AST (ast.rs). Pure,
//  dependency-free. A direct port of parse.js: same tokenizer (// line
//  comments, string/number/identifier rules, PUNCT set), same
//  recursive-descent grammar, same 1-based line tracking, same error
//  messages. Forgiving about whitespace/comments, strict about structure.
// =====================================================================

use crate::ast::*;
use crate::error::ParseError;

type PResult<T> = Result<T, ParseError>;

// ---- tokenizer -------------------------------------------------------

#[derive(Debug, Clone, PartialEq)]
enum Tok {
    Punct(char), // one of { } ( ) , : =
    Str(String),
    Num(f64),
    Id(String),
    Eof,
}

#[derive(Debug, Clone)]
struct Token {
    t: Tok,
    line: u32,
}

fn is_punct(c: char) -> bool {
    matches!(c, '{' | '}' | '(' | ')' | ',' | ':' | '=')
}

fn tokenize(src: &str) -> PResult<Vec<Token>> {
    let chars: Vec<char> = src.chars().collect();
    let n = chars.len();
    let mut toks: Vec<Token> = Vec::new();
    let mut line: u32 = 1;
    let mut i = 0usize;

    while i < n {
        let c = chars[i];
        if c == '\n' {
            line += 1;
            i += 1;
            continue;
        }
        if c == ' ' || c == '\t' || c == '\r' {
            i += 1;
            continue;
        }
        // line comment: // to end of line
        if c == '/' && i + 1 < n && chars[i + 1] == '/' {
            while i < n && chars[i] != '\n' {
                i += 1;
            }
            continue;
        }
        if is_punct(c) {
            toks.push(Token { t: Tok::Punct(c), line });
            i += 1;
            continue;
        }
        // string literal
        if c == '"' {
            let mut j = i + 1;
            let mut s = String::new();
            while j < n && chars[j] != '"' {
                s.push(chars[j]);
                j += 1;
            }
            if j >= n {
                return Err(ParseError::new("unterminated string", Some(line)));
            }
            toks.push(Token { t: Tok::Str(s), line });
            i = j + 1;
            continue;
        }
        // number: digit, or '-' followed by a digit
        if c.is_ascii_digit() || (c == '-' && i + 1 < n && chars[i + 1].is_ascii_digit()) {
            let mut j = i + 1;
            while j < n && matches!(chars[j], '0'..='9' | '.' | 'e' | 'E' | '+' | '-') {
                j += 1;
            }
            let raw: String = chars[i..j].iter().collect();
            match raw.parse::<f64>() {
                Ok(v) => toks.push(Token { t: Tok::Num(v), line }),
                Err(_) => return Err(ParseError::new(format!("bad number \"{}\"", raw), Some(line))),
            }
            i = j;
            continue;
        }
        // identifier / keyword: [A-Za-z_][A-Za-z0-9_.]*
        if c.is_ascii_alphabetic() || c == '_' {
            let mut j = i + 1;
            while j < n && (chars[j].is_ascii_alphanumeric() || chars[j] == '_' || chars[j] == '.') {
                j += 1;
            }
            let s: String = chars[i..j].iter().collect();
            toks.push(Token { t: Tok::Id(s), line });
            i = j;
            continue;
        }
        return Err(ParseError::new(format!("unexpected character \"{}\"", c), Some(line)));
    }
    toks.push(Token { t: Tok::Eof, line });
    Ok(toks)
}

// ---- recursive-descent parser ---------------------------------------

struct Parser {
    toks: Vec<Token>,
    p: usize,
}

/// Human-facing token name, matching the JS `tk.v ?? tk.t` reporting.
fn tok_name(t: &Tok) -> String {
    match t {
        Tok::Punct(c) => c.to_string(),
        Tok::Str(s) => s.clone(),
        Tok::Num(v) => fmt_num(*v),
        Tok::Id(s) => s.clone(),
        Tok::Eof => "eof".to_string(),
    }
}

/// Format a number the way JS String() would for the common integer/float
/// cases that appear in these scripts (used only in error messages).
fn fmt_num(v: f64) -> String {
    if v.fract() == 0.0 && v.is_finite() {
        format!("{}", v as i64)
    } else {
        format!("{}", v)
    }
}

impl Parser {
    fn new(toks: Vec<Token>) -> Self {
        Parser { toks, p: 0 }
    }
    fn peek(&self) -> &Token {
        &self.toks[self.p]
    }
    fn next(&mut self) -> Token {
        let t = self.toks[self.p].clone();
        self.p += 1;
        t
    }
    fn at_punct(&self, c: char) -> bool {
        matches!(&self.peek().t, Tok::Punct(x) if *x == c)
    }
    fn at_eof(&self) -> bool {
        matches!(self.peek().t, Tok::Eof)
    }
    fn at_kw(&self, kw: &str) -> bool {
        matches!(&self.peek().t, Tok::Id(v) if v == kw)
    }
    fn eat_punct(&mut self, c: char) -> PResult<()> {
        let tk = self.peek().clone();
        if !self.at_punct(c) {
            return Err(ParseError::new(
                format!("expected \"{}\", got \"{}\"", c, tok_name(&tk.t)),
                Some(tk.line),
            ));
        }
        self.next();
        Ok(())
    }
    fn eat_kw(&mut self, kw: &str) -> PResult<()> {
        let tk = self.peek().clone();
        if !self.at_kw(kw) {
            return Err(ParseError::new(
                format!("expected \"{}\", got \"{}\"", kw, tok_name(&tk.t)),
                Some(tk.line),
            ));
        }
        self.next();
        Ok(())
    }
    fn id(&mut self) -> PResult<String> {
        let tk = self.peek().clone();
        match &tk.t {
            Tok::Id(v) => {
                let v = v.clone();
                self.next();
                Ok(v)
            }
            _ => Err(ParseError::new(
                format!("expected identifier, got \"{}\"", tok_name(&tk.t)),
                Some(tk.line),
            )),
        }
    }
    fn num(&mut self) -> PResult<f64> {
        let tk = self.peek().clone();
        match tk.t {
            Tok::Num(v) => {
                self.next();
                Ok(v)
            }
            _ => Err(ParseError::new(
                format!("expected number, got \"{}\"", tok_name(&tk.t)),
                Some(tk.line),
            )),
        }
    }

    // ---- top level ----
    fn parse_program(&mut self) -> PResult<Spec> {
        let tk = self.peek().clone();
        if self.at_kw("agent") {
            return Ok(Spec::Agent(self.parse_agent()?));
        }
        if self.at_kw("society") {
            return Ok(Spec::Society(self.parse_society()?));
        }
        Err(ParseError::new(
            format!("expected \"agent\" or \"society\", got \"{}\"", tok_name(&tk.t)),
            Some(tk.line),
        ))
    }

    // ---- agent ----
    fn parse_agent(&mut self) -> PResult<AgentSpec> {
        let line = self.peek().line;
        self.eat_kw("agent")?;
        let name = self.id()?;
        self.eat_punct('{')?;
        let mut spec = AgentSpec {
            name,
            line,
            purpose: None,
            scenes: Vec::new(),
            self_graph: None,
            budget: None,
            floor: None,
            coherence: None,
        };
        while !self.at_punct('}') {
            if self.at_eof() {
                return Err(ParseError::new("unclosed \"agent\" block", Some(line)));
            }
            if self.at_kw("purpose") {
                spec.purpose = Some(self.parse_purpose()?);
            } else if self.at_kw("scenes") {
                spec.scenes = self.parse_scenes()?;
            } else if self.at_kw("self") {
                spec.self_graph = Some(self.parse_self()?);
            } else if self.at_kw("budget") {
                self.eat_kw("budget")?;
                spec.budget = Some(self.num()?);
            } else if self.at_kw("floor") {
                self.eat_kw("floor")?;
                spec.floor = Some(self.num()?);
            } else if self.at_kw("coherence") {
                spec.coherence = Some(self.parse_coherence()?);
            } else {
                let bad = self.peek().clone();
                return Err(ParseError::new(
                    format!("unexpected \"{}\" in agent body", tok_name(&bad.t)),
                    Some(bad.line),
                ));
            }
        }
        self.eat_punct('}')?;
        Ok(spec)
    }

    fn parse_purpose(&mut self) -> PResult<Purpose> {
        self.eat_kw("purpose")?;
        let tk = self.peek().clone();
        if self.at_kw("minimise") || self.at_kw("minimize") {
            self.next();
            let potential = self.id()?;
            return Ok(Purpose::Minimise { potential, line: tk.line });
        }
        if self.at_kw("reach") {
            self.next();
            let outcome = self.id()?;
            return Ok(Purpose::Reach { outcome, line: tk.line });
        }
        Err(ParseError::new(
            "purpose must be \"minimise <phi>\" or \"reach <outcome>\"",
            Some(tk.line),
        ))
    }

    fn parse_scenes(&mut self) -> PResult<Vec<Scene>> {
        self.eat_kw("scenes")?;
        self.eat_punct('{')?;
        let mut scenes = Vec::new();
        while !self.at_punct('}') {
            if self.at_eof() {
                return Err(ParseError::new("unclosed \"scenes\" block", Some(self.peek().line)));
            }
            let line = self.peek().line;
            self.eat_kw("scene")?;
            let name = self.id()?;
            self.eat_kw("serves")?;
            let serves = self.id()?;
            self.eat_kw("with")?;
            let hook = self.id()?;
            scenes.push(Scene { name, serves, hook, line });
        }
        self.eat_punct('}')?;
        Ok(scenes)
    }

    fn parse_self(&mut self) -> PResult<SelfGraph> {
        let line = self.peek().line;
        self.eat_kw("self")?;
        self.eat_punct('{')?;
        let mut parts: Vec<String> = Vec::new();
        let mut separations: Vec<Separation> = Vec::new();
        while !self.at_punct('}') {
            if self.at_eof() {
                return Err(ParseError::new("unclosed \"self\" block", Some(line)));
            }
            if self.at_kw("parts") {
                self.eat_kw("parts")?;
                parts = self.parse_id_list()?;
            } else if self.at_kw("separations") {
                self.eat_kw("separations")?;
                separations = self.parse_separations()?;
            } else {
                let bad = self.peek().clone();
                return Err(ParseError::new(
                    format!("unexpected \"{}\" in self body", tok_name(&bad.t)),
                    Some(bad.line),
                ));
            }
        }
        self.eat_punct('}')?;
        Ok(SelfGraph { parts, separations, line })
    }

    // { a, b, c }
    fn parse_id_list(&mut self) -> PResult<Vec<String>> {
        self.eat_punct('{')?;
        let mut ids = Vec::new();
        while !self.at_punct('}') {
            ids.push(self.id()?);
            if self.at_punct(',') {
                self.next();
            } else if !self.at_punct('}') {
                return Err(ParseError::new("expected \",\" or \"}\"", Some(self.peek().line)));
            }
        }
        self.eat_punct('}')?;
        Ok(ids)
    }

    // { (a, b: 3), (b, c: 2), ... }
    fn parse_separations(&mut self) -> PResult<Vec<Separation>> {
        self.eat_punct('{')?;
        let mut seps = Vec::new();
        while !self.at_punct('}') {
            let line = self.peek().line;
            self.eat_punct('(')?;
            let a = self.id()?;
            self.eat_punct(',')?;
            let b = self.id()?;
            self.eat_punct(':')?;
            let cost = self.num()?;
            self.eat_punct(')')?;
            seps.push(Separation { a, b, cost, line });
            if self.at_punct(',') {
                self.next();
            } else if !self.at_punct('}') {
                return Err(ParseError::new("expected \",\" or \"}\"", Some(self.peek().line)));
            }
        }
        self.eat_punct('}')?;
        Ok(seps)
    }

    fn parse_coherence(&mut self) -> PResult<Coherence> {
        let line = self.peek().line;
        self.eat_kw("coherence")?;
        // optional "keeps"
        if self.at_kw("keeps") {
            self.next();
        }
        let keeps = self.parse_id_list()?;
        Ok(Coherence { keeps, line })
    }

    // ---- society ----
    fn parse_society(&mut self) -> PResult<SocietySpec> {
        let line = self.peek().line;
        self.eat_kw("society")?;
        let name = self.id()?;
        self.eat_punct('{')?;
        let mut spec = SocietySpec {
            name,
            line,
            members: Vec::new(),
            ties: Vec::new(),
            couple: None,
        };
        while !self.at_punct('}') {
            if self.at_eof() {
                return Err(ParseError::new("unclosed \"society\" block", Some(line)));
            }
            if self.at_kw("agent") {
                spec.members.push(Member::Agent(self.parse_agent()?));
            } else if self.at_kw("tie") {
                self.eat_kw("tie")?;
                self.eat_punct('(')?;
                let a = self.id()?;
                self.eat_punct(',')?;
                let b = self.id()?;
                self.eat_punct(':')?;
                let cost = self.num()?;
                self.eat_punct(')')?;
                spec.ties.push(Tie { a, b, cost });
            } else if self.at_kw("couple") {
                self.eat_kw("couple")?;
                spec.couple = Some(self.num()?);
            } else {
                // a bare member reference by name
                let bad = self.peek().clone();
                match &bad.t {
                    Tok::Id(_) => {
                        let name = self.id()?;
                        spec.members.push(Member::Ref { name, line: bad.line });
                    }
                    _ => {
                        return Err(ParseError::new(
                            format!("unexpected \"{}\" in society body", tok_name(&bad.t)),
                            Some(bad.line),
                        ));
                    }
                }
            }
        }
        self.eat_punct('}')?;
        Ok(spec)
    }
}

/// Parse an Agent Smith script into a spec AST.
pub fn parse(src: &str) -> PResult<Spec> {
    let toks = tokenize(src)?;
    let mut parser = Parser::new(toks);
    let spec = parser.parse_program()?;
    if !parser.at_eof() {
        let tk = parser.peek().clone();
        return Err(ParseError::new(
            format!("unexpected trailing \"{}\"", tok_name(&tk.t)),
            Some(tk.line),
        ));
    }
    Ok(spec)
}
